rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,467 @@
+        // Attempt to read the data from the inner body, then update the
+        // throughput logs.
+        let mut this = self.as_mut().project();
+        let poll_res = match this.inner.poll_trailers(cx) {
+            Poll::Ready(Ok(Some(trailers))) => {
+                let size_of_all_keys: u64 = trailers.keys().map(|k| k.as_str().len() as u64).sum();
+                let size_of_all_values: u64 = trailers.values().map(|v| v.len() as u64).sum();
+                this.throughput_logs
+                    .push((now, size_of_all_keys + size_of_all_values));
+                a_log_was_pushed_this_poll = true;
+                Poll::Ready(Ok(Some(trailers)))
+            }
+            Poll::Pending => Poll::Pending,
+            // If we've read all the data or an error occurred, then return that result.
+            res => return res,
+        };
+
+        // Check the sleep future to see if it needs refreshing.
+        let mut sleep_fut = this.sleep_fut.take().unwrap_or_else(|| {
+            this.async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed())
+        });
+        // The strategy here is the same as when polling for data. Read those
+        // comments first if you have questions.
+        if let Poll::Ready(()) = pin!(&mut sleep_fut).poll(cx) {
+            sleep_fut = this
+                .async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed());
+            if !a_log_was_pushed_this_poll {
+                this.throughput_logs.set_floating_back(now);
+            }
+            cx.waker().wake_by_ref();
+        };
+        this.sleep_fut.replace(sleep_fut);
+
+        // Calculate the current throughput and emit an error if it's too low.
+        let actual_throughput = this.throughput_logs.calculate_throughput()?;
+        let is_below_minimum_throughput = actual_throughput
+            .map(|t| t < self.minimum_throughput)
+            .unwrap_or_default();
+        if is_below_minimum_throughput {
+            Poll::Ready(Err(Box::new(Error::ThroughputBelowMinimum {
+                expected: self.minimum_throughput,
+                actual: actual_throughput.unwrap(),
+            })))
+        } else {
+            poll_res
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Error {
+    ThroughputBelowMinimum {
+        expected: Throughput,
+        actual: Throughput,
+    },
+    TimeTravel(BoxError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::ThroughputBelowMinimum { expected, actual } => {
+                write!(
+                    f,
+                    "minimum throughput was specified at {expected}, but throughput of {actual} was observed",
+                )
+            }
+            Self::TimeTravel(_) => write!(
+                f,
+                "negative time has elapsed while reading the inner body, this is a bug"
```

Comment:
it isn't really though—we're using SystemTime and not `Instant`. SystemTime doesn't guarantee monotonicity:
> Distinct from the Instant type, this time measurement is not monotonic. This means that you can save a file to the file system, then save another file to the file system, and the second file has a SystemTime measurement earlier than the first. In other words, an operation that happens after another operation in real time may have an earlier SystemTime!

https://doc.rust-lang.org/std/time/struct.SystemTime.html

We should probably try to handle this gracefully or add some sort of `InstantTimeSource` in our bag of tricks

Reply:
Agreed, just not in this PR.